File: Days 41 - 50/Day 46/LongestPalindromicSubstring.cpp
#include <algorithm>
#include <iostream>
#include <string>

bool IsPalindrome(const std::string& string)
{
	std::string reversedString = string;
	std::reverse_copy(string.begin(), string.end(), reversedString.begin());

	return string == reversedString;
}

std::string GetLongestPalindromicSubstring(const std::string& string)
{
	if (string.empty() || IsPalindrome(string))
	{
		return string;
	}

	std::string withoutFirst = GetLongestPalindromicSubstring(string.substr(1));
	std::string withoutLast = GetLongestPalindromicSubstring(string.substr(0, string.length() - 1));

	return withoutFirst.length() >= withoutLast.length() ? withoutFirst : withoutLast;
}

int main(int argc, const char* argv[])
{
	std::cout << GetLongestPalindromicSubstring("aabcdcb") << "\n";
	std::cout << GetLongestPalindromicSubstring("bananas") << "\n";

	std::cin.get();

	return 0;
}

File: Days 41 - 50/Day 46/LongestPalindromicSubstring.cs
using System;

namespace DailyCodingProblem
{
	internal class Day46
	{
		private static int Main(string[] args)
		{
			Console.WriteLine(GetLongestPalindromicSubstring("aabcdcb"));
			Console.WriteLine(GetLongestPalindromicSubstring("bananas"));

			Console.ReadLine();

			return 0;
		}

		private static string GetLongestPalindromicSubstring(string text)
		{
			if (text.Length == 0 || IsPalindrome(text))
			{
				return text;
			}

			string withoutFirst = GetLongestPalindromicSubstring(text.Substring(1));
			string withoutLast = GetLongestPalindromicSubstring(text.Substring(0, text.Length - 1));

			return withoutFirst.Length >= withoutLast.Length ? withoutFirst : withoutLast;
		}

		private static bool IsPalindrome(string text)
		{
			char[] chars = text.ToCharArray();
			Array.Reverse(chars);

			return text == new string(chars);
		}
	}
}